deviceplugin/deviceplugin.go

```diff
@@ -0,0 +1,144 @@
+	"path"
+	"time"
+)
+
+type DevicePlugin struct {
+	socket       string
+	resourceName string
+	stop         chan interface{}
+	server       *grpc.Server
+}
+
+func NewDevicePlugin(serversock string, resourcename string) *DevicePlugin {
+	return &DevicePlugin{
+		socket:       serversock,
+		stop:         make(chan interface{}),
+		resourceName: resourcename,
+	}
+}
+
+func dial(timeoutContext context.Context, unixSocketPath string) (*grpc.ClientConn, error) {
+	return grpc.DialContext(
+		timeoutContext,
+		unixSocketPath,
+		grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+}
+
+func (d *DevicePlugin) cleanup() error {
+	if err := os.Remove(d.socket); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	return nil
+}
+
+func (d *DevicePlugin) Start() error {
+	err := d.cleanup()
+
+	if err != nil {
+		return err
+	}
+
+	sock, err := net.Listen("unix", d.socket)
+	if err != nil {
+		return err
+	}
+	d.server = grpc.NewServer([]grpc.ServerOption{}...)
+	pluginapi.RegisterDevicePluginServer(d.server, d)
+
+	go d.server.Serve(sock)
+
+	ctx := context.TODO()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	conn, err := dial(ctx, d.socket)
+	if err != nil {
+		return err
+	}
+	conn.Close()
+
+	return nil
+}
+
+func (d *DevicePlugin) Stop() error {
+	if d.server == nil {
+		return nil
+	}
+
+	d.server.Stop()
+	d.server = nil
+	close(d.stop)
+	return d.cleanup()
+}
+
+func (d *DevicePlugin) Serve() error {
+	err := d.Start()
+	if err != nil {
+		log.Printf("Could not start device plugin %s", err)
+	}
+	log.Println("Starting to serve on", d.socket)
+
+	err = d.Register(pluginapi.KubeletSocket, d.resourceName)
+	if err != nil {
+		log.Printf("Could not register device plugin %s", err)
+		return err
+	}
+	log.Println("Registered device plugin with Kubelet")
+	return nil
+}
+
+// Define functions needed to meet the Kubernetes DevicePlugin API
+
+func (d *DevicePlugin) GetDevicePluginOptions(context.Context, *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	return &pluginapi.DevicePluginOptions{}, nil
+}
+
+func (d *DevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	return nil, nil
+}
+
+func (d *DevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+	return nil
+}
+
+func (d *DevicePlugin) PreStartContainer(context.Context, *pluginapi.PreStartContainerRequest) (*pluginapi.PreStartContainerResponse, error) {
+	return &pluginapi.PreStartContainerResponse{}, nil
+}
+
+func (d *DevicePlugin) Register(kubeletEndpoint, resourceName string) error {
+	ctx := context.TODO()
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	conn, err := dial(ctx, kubeletEndpoint)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
```

Comment:
Defer is a stack, so we should make sure we defer conn.Close() before we defer cancel()

Reply:
I'm not finding where I defer cancel...